### PR TITLE
SC: Add an old anchoring tx handling for compatibility

### PR DESCRIPTION
**Pull request by jeonghonah**
## Proposed changes

- To handle previous anchoring tx type gracefully, added an exception handling for RLP decoding failure of an anchoring data.
- If RPL decoding failed with AnchoringDataInternalType0, it try to decoding with AnchoringDataInternalNoType that is an old anchoring tx type before merging #176.
- Added a related unit test.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- #176

### CODE REVIEW

File: node/sc/sub_bridge_handler.go [hunk 1]
```diff
@@ -310,43 +315,53 @@ func (sbh *SubBridgeHandler) broadcastServiceChainTx() {
...
 
+// decodeAnchoringTx decodes an anchoring transaction.
+func (sbh *SubBridgeHandler) decodeAnchoringTx(data []byte) (common.Hash, *big.Int, error) {
+	anchoringData := new(types.AnchoringData)
+	if err := rlp.DecodeBytes(data, anchoringData); err != nil {
+		anchoringDataNoType := new(types.AnchoringDataInternalNoType)
+		if err := rlp.DecodeBytes(data, anchoringDataNoType); err != nil {
+			return common.Hash{}, nil, err
+		}
+		logger.Trace("decoded type0 anchoring tx", "blockNum", anchoringDataNoType.BlockNumber.String(), "blockHash", anchoringDataNoType.BlockHash.String(), "txHash", anchoringDataNoType.TxHash.String())
```
Comment (aidan-kwon) [inline node/sc/sub_bridge_handler.go]: The data is decoded as `types.AnchoringDataInternalNoType` type, but the log says "decoded type0 anchoring tx".

File: blockchain/types/anchoring_data.go [hunk 2]
```diff
@@ -31,6 +31,16 @@ type AnchoringData struct {
 	Data []byte
 }
 
+// AnchoringDataInternalNoType is an old anchoring type that does not support an data type.
+type AnchoringDataInternalNoType struct {
```
Comment (aidan-kwon) [inline blockchain/types/anchoring_data.go]: How changing the name to `AnchoringDataInternalDefault` if it is used as a fallback type.

Comment (jeonghonah) [reply blockchain/types/anchoring_data.go]: Actually, the default is type0.

File: node/sc/sub_bridge_handler.go (hunk 1 above)
Comment (jeonghonah) [reply node/sc/sub_bridge_handler.go]: Fixed.

File: node/sc/main_event_handler.go [hunk 3]
```diff
@@ -80,13 +80,42 @@ func (mce *MainChainEventHandler) ConvertChildChainBlockHashToParentChainTxHash(
...
+	anchoringData := new(types.AnchoringData)
+	if err := rlp.DecodeBytes(data, anchoringData); err != nil {
+		// Try to decode old type without a type support for compatibility.
+		anchoringDataLegacy := new(types.AnchoringDataLegacy)
+		if err := rlp.DecodeBytes(data, anchoringDataLegacy); err != nil {
+			logger.Error("writeChildChainTxHashFromBlock : failed to decode anchoring data", "txHash", tx.Hash().String())
+			return
+		}
+		mce.mainbridge.chainDB.WriteChildChainTxHash(anchoringDataLegacy.BlockHash, tx.Hash())
+		logger.Trace("Write type0 anchoring data on chainDB", "blockHash", anchoringDataLegacy.BlockHash.String(), "txHash", tx.Hash().String())
```
Comment (aidan-kwon) [inline node/sc/main_event_handler.go]:
```suggestion
		logger.Trace("Write legacy anchoring data on chainDB", "blockHash", anchoringDataLegacy.BlockHash.String(), "txHash", tx.Hash().String())
```

File: node/sc/main_event_handler.go [hunk 4]
```diff
@@ -80,13 +80,42 @@ func (mce *MainChainEventHandler) ConvertChildChainBlockHashToParentChainTxHash(
...
+		return
+	}
+	if anchoringData.Type == types.AnchoringDataType0 {
+		anchoringDataInternal := new(types.AnchoringDataInternalType0)
+		if err := rlp.DecodeBytes(anchoringData.Data, anchoringDataInternal); err != nil {
+			logger.Error("writeChildChainTxHashFromBlock : failed to decode anchoring data", "txHash", tx.Hash().String())
+			return
+		}
+		mce.mainbridge.chainDB.WriteChildChainTxHash(anchoringDataInternal.BlockHash, tx.Hash())
+		logger.Trace("Write type1 anchoring data on chainDB", "blockHash", anchoringDataInternal.BlockHash.String(), "txHash", tx.Hash().String())
```
Comment (aidan-kwon) [inline node/sc/main_event_handler.go]:
```suggestion
		logger.Trace("Write type0 anchoring data on chainDB", "blockHash", anchoringDataInternal.BlockHash.String(), "txHash", tx.Hash().String())
```

File: node/sc/main_event_handler.go [hunk 5]
```diff
@@ -80,13 +80,42 @@ func (mce *MainChainEventHandler) ConvertChildChainBlockHashToParentChainTxHash(
...
+	if err != nil {
+		logger.Error("writeChildChainTxHashFromBlock : failed to get anchoring data from the tx", "txHash", tx.Hash().String())
+		return
+	}
+	anchoringData := new(types.AnchoringData)
+	if err := rlp.DecodeBytes(data, anchoringData); err != nil {
+		// Try to decode old type without a type support for compatibility.
+		anchoringDataLegacy := new(types.AnchoringDataLegacy)
+		if err := rlp.DecodeBytes(data, anchoringDataLegacy); err != nil {
+			logger.Error("writeChildChainTxHashFromBlock : failed to decode anchoring data", "txHash", tx.Hash().String())
```
Comment (jimni1222) [inline node/sc/main_event_handler.go]: How about change error log message? 
writeChildChainTxHashFromBlock is another function, ;-; L107, L113 too

Comment (jeonghonah) [reply node/sc/main_event_handler.go]: I agree with you but we delete this method in future PR since it is not used anymore.

**Pull request merged**